### ISSUE TITLE
ARRAY_PUSH action with Immutable give Invalid keyPath error

### ISSUE THREAD

**Issue by JonatanGarciaClavo** [erikras/redux-form#2774]
I think this is a bug related with validate function and immutable, because if I don't pass validate function ARRAY_PUSH actions works fine.

### What is the current behavior?
![screen shot 2017-04-04 at 16 04 40](https://cloud.githubusercontent.com/assets/4894569/24661372/83c9f82e-1952-11e7-9502-fa8a8cbe1842.png)

At the moment I try to push on field array new data. In this case I am trying to initialize new condition with a default operator. So idea of final result in plain JS will be:
```js
{
  title: 'title',
  conditions: [
    {
       attribute: 'bla',
       operation: '===',
       value: 'true',
       operator: 'AND',
    },
    {
       attribute: 'bla',
       operation: '===',
       value: 'true',
       operator: 'AND',
    },
....
  ],
}
```

### What is the expected behavior?
I expect when I try to add new data to list of conditions doesn't crash

### What's your environment?
redux-form: 6.6.1
redux: 3.6.0
redux-immutable: 4.0.0
immutable: 4.0.0-rc.2
material-ui: 0.17.0

Let me know if you need more info about it

### Other informations
After some research I "fix" it if I add fromJs to line 390 on reducer.js file
result = setIn(result, 'syncErrors', fromJS(syncErrors))
I detect that state store syncErrors as Immutable.Map but all inside of it is plain object so when you use getIn and want to access deeper this error shows because deeper key is plain object not Immutable data. So thats it is reason I add that fromJS, but after run tests I saw I break 7 tests so I have feeling that maybe it isn't real problem.

This is killing me because I can't continue until this is fixed, let me know how I could help you or if you need anything from my side.

**Comment by erikras** [comment]
Yep, this is a bug. I've recreated it in a test. Thanks!

**Comment by erikras** [comment]
No, I take that back. It seems to be working as I expect. Look at the test in https://github.com/erikras/redux-form/commit/946b2ba1f4e78f46b382d21fa58ea9d347837db8.

I'm going to need to see your validation function. Or some minimal version of it that continues to generate this error.

**Comment by JonatanGarciaClavo** [comment]
```
// validations.js

export const required = (value) => (
  value !== undefined && value !== null && (!isEmpty(value) || isNumber(value))
);

// constans.js
export const FORM_KEY_COMMNAME = 'commName';
export const FORM_KEY_EVENT = 'event';
export const FORM_KEY_CANCEL_EVENT = 'cancelEvent';
export const FORM_KEY_CONDITIONS = 'conditions';
export const FORM_KEY_CONDITION_ATTRIBUTE = 'attribute';
export const FORM_KEY_CONDITION_OPERATION = 'operation';
export const FORM_KEY_CONDITION_VALUE = 'value';
export const FORM_KEY_CONDITION_OPERATOR = 'operator';


import { List } from 'immutable';
import {
  required,
  // isJSONArrayOfNumbers,
  // isJSONArrayOfStrings,
  // isStringOperator,
} from '../../utils/form/validations';
import {
  FORM_KEY_COMMNAME,
  FORM_KEY_EVENT,
  // FORM_KEY_CANCEL_EVENT,
  FORM_KEY_CONDITIONS,
  FORM_KEY_CONDITION_ATTRIBUTE,
  FORM_KEY_CONDITION_OPERATION,
  FORM_KEY_CONDITION_VALUE,
} from './constants';
import messages from './messages';

export default function validate(values) {
  const errors = {};
  if (!required(values.get(FORM_KEY_COMMNAME))) {
    errors[FORM_KEY_COMMNAME] = messages.titleRequired.defaultMessage;
  }
  if (!required(values.get(FORM_KEY_EVENT))) {
    errors[FORM_KEY_EVENT] = messages.eventRequired.defaultMessage;
  }
  // if (!required(values.get('FORM_KEY_CANCEL_EVENT'))) {
  //   errors[FORM_KEY_CANCEL_EVENT] = messages.cancelEventRequired.defaultMessage;
  // }
  const conditionsArrayErrors = [];
  values.get(FORM_KEY_CONDITIONS, new List()).forEach((condition, index) => {
    const conditionsErrors = {};
    if (!required(condition.get(FORM_KEY_CONDITION_ATTRIBUTE))) {
      conditionsErrors[FORM_KEY_CONDITION_ATTRIBUTE] = messages.attributeRequired.defaultMessage;
      conditionsArrayErrors[index] = conditionsErrors;
    }
    if (!required(condition.get(FORM_KEY_CONDITION_OPERATION))) {
      conditionsErrors[FORM_KEY_CONDITION_OPERATION] = messages.operationRequired.defaultMessage;
      conditionsArrayErrors[index] = conditionsErrors;
    }
    if (!required(condition.get(FORM_KEY_CONDITION_VALUE))) {
      conditionsErrors[FORM_KEY_CONDITION_VALUE] = messages.valueRequired.defaultMessage;
      conditionsArrayErrors[index] = conditionsErrors;
    }
  });
  if (conditionsArrayErrors.length) {
    errors[FORM_KEY_CONDITIONS] = conditionsArrayErrors;
  }
  return errors;
}

```

Next extracted code it is part of Field Array component prop
````
renderConditionFields({ fields, selectedEvent, attrOptions, operationOptions }) {
    console.log('selectedEvent', selectedEvent);
    return (
      <div>
        {fields.map((condition, index, { get }) => {
          console.log('renderConditionFields get', get());
          const conditionValue = get().get(index) || new Map();
          const attr = conditionValue.get('attribute');
          const attrFieldOptions = selectedEvent ? attrOptions[selectedEvent.eventId] || [] : [];
          const operationFieldOptions = attr ? operationOptions[selectedEvent.jsonSchema.properties[attr].type] : [];
          return (
            <div key={index}>
              <div style={{ display: 'flex', flex: '0 0 auto', flexDirection: 'column', background: 'white' }}>
                <Field
                  name={`${condition}.${FORM_KEY_CONDITION_ATTRIBUTE}`}
                  component={({ input, ...others }) => {
                    const newInput = {
                      ...input,
                      onChange: (...opts) => {
                        this.props.change(`${condition}`, new Map({
                          [FORM_KEY_CONDITION_OPERATOR]: 'AND',
                        }));
                        input.onChange(...opts);
                      },
                    };
                    return this.renderSelectField({ input: newInput, ...others });
                  }}
                  label="Attribute"
                  disabled={!selectedEvent}
                  options={attrFieldOptions}
                />
                <Field
                  name={`${condition}.${FORM_KEY_CONDITION_OPERATION}`}
                  component={this.renderSelectField}
                  label="Operation"
                  disabled={!conditionValue.get(FORM_KEY_CONDITION_ATTRIBUTE)}
                  options={operationFieldOptions}
                />
                <Field
                  name={`${condition}.${FORM_KEY_CONDITION_VALUE}`}
                  component={this.renderTextField}
                  label="Value"
                  disabled={!conditionValue.get(FORM_KEY_CONDITION_OPERATION)}
                />
              </div>
              {(fields.length - 1) !== index ?
                <div>
                  <Field
                    name={`${condition}.${FORM_KEY_CONDITION_OPERATOR}`}
                    component={this.renderRadioGroup}
                  >
                    <RadioButton value="OR" label="OR" />
                    <RadioButton value="AND" label="AND" />
                  </Field>
                </div>
              : null}
              {(fields.length - 1) !== 0 ? <div style={{ display: 'flex', flex: '0 0 auto', flexDirection: 'column' }}>
                <span
                  style={{ padding: '1em' }}
                  onClick={() => fields.remove(index)}
                >Remove
                </span>
              </div>
              : null}
            </div>
          );
        })}
        <div style={{ margin: '1.5em' }}>
          <span onClick={() => fields.push(new Map({ [FORM_KEY_CONDITION_OPERATOR]: 'AND' }))}>+ New condition</span>
        </div>
      </div>
    );
  }
```

Let me know if you need something else.
Pd.- sorry about dirty code but I was testing first if it is posible to do what I want then I will refactor it to be more reusable for multiple forms I will need.
Regards.

**Comment by JonatanGarciaClavo** [comment]
I reproduce it here [Material-ui example](https://github.com/JonatanGarciaClavo/redux-form/tree/master/examples/material-ui)
I hope with this info you could tell me what I am doing wrong or it is at the end a bug.
Regards.

**Comment by erikras** [comment]
I've managed to figure out that the reason my test is passing is due to a very strange inconsistency in the behavior on Node and in the browser. I have reported it to them here: https://github.com/facebook/immutable-js/issues/1201

Apart from that, this is a bug in my library that I need to fix.

**Comment by rahulthewall** [comment]
Hi,

I believe I have a similar issue. Here's a snippet of the form that might be relevant:

```
<Field
	className={s.projectName}
	type="text"
	name="realEstate.project-name"
	label={t('object.form.projectName')}
	updateDraft={value => update(['project-name'], value)}
	component={InputField}
/>
<Field
	name="realEstate.short-description"
	label={t('object.form.projectShortDescription')}
	updateDraft={value => update(['short-description'], value)}
	component={TextareaField}
/>

...

export default localize(reduxForm({
	form: 'dealEstate.realEstate.editor',
	enableReinitialize: false,
	validate: validateForm,
})(withStyles(s)(RealEstateForm)));
```

If I use the form with `enableReinitialize` set to `false`, I see the following errors on my console:

```
immutable.js:872 Uncaught Error: invalid keyPath
    at invariant (immutable.js:872)
    at updateInDeepMap (immutable.js:1976)
    at updateInDeepMap (immutable.js:1982)
    at updateInDeepMap (immutable.js:1982)
    at Map.updateIn (immutable.js:1280)
    at Map.deleteIn (immutable.js:1266)
    at deleteIn (index.js:46)
    at deleteInWithCleanUp (deleteInWithCleanUp.js:29)
    at reducer.js:153
    at reducer (reducer.js:463)
```

As soon as I set `enableReinitialize` to `true`, I don't see these errors. However, I don't want to use `enableReinitialize` to `true`, because I want to use the `touched` values in the fields which gets reset when the form is re-initialised.

**Comment by JonatanGarciaClavo** [comment]
@rahulthewall I have `enableReinitialize` to `true` and it still happening, so I guess it is something else, let see if @erikras has time to check it

**Comment by JonatanGarciaClavo** [comment]
@erikras just in case that helps, but I think it is problem with immutable 4 rc, because I saw some issues related to that problem.

**Comment by rahulthewall** [comment]
I am using immutable-3.8.1 though.

**Comment by JonatanGarciaClavo** [comment]
@rahulthewall then i have no idea xD

**Comment by erikras** [comment]
@rahulthewall Your problem could very well be related, if it has to do with sync errors. `redux-form` was not doing a good job of reading and writing the sync errors as a plain js object. #2779 might fix your problem as well.

**Comment by JonatanGarciaClavo** [comment]
@erikras when it will be available to be updated via npm?

**Comment by rahulthewall** [comment]
@erikras I pulled from git, did an `npm i && npm build` and then copied the folders over to my `node_modules` folder. I still get the error. Is the right way to check whether my problem was fixed or not?

**Comment by erikras** [comment]
Fix published in [`v6.6.2`](https://github.com/erikras/redux-form/releases/tag/v6.6.2).
